### PR TITLE
HIVE-23389

### PR DESCRIPTION
**Pull request by jcamachor**

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/Bug.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.optimizer.calcite;
+
+import org.apache.calcite.util.Util;
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/Bug.java:20]: Useless import?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/Bug.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * Holder for a list of constants describing which bugs which have not been
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/Bug.java:23]: typo: **which** bugs **which**
suggestion: A holder of constants describing existing bugs in Calcite.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterMergeRule.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            topProgram,
+            bottomProgram,
+            rexBuilder);
+
+    RexNode newCondition = expandLocalRef(rexBuilder,
+        mergedProgram.getCondition(), mergedProgram.getExprList());
+
+    final RelBuilder relBuilder = call.builder();
+    relBuilder.push(bottomFilter.getInput())
+        .filter(newCondition);
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterMergeRule.java:76]: I have the impression that we could replace this code with:
```
final RelBuilder relBuilder = call.builder();
    relBuilder.push(bottomFilter.getInput())
        .filter(topFilter.getCondition(), bottomFilter.getCondition());
```
If I remember well, there is a simplification inside the `builder.filter(...)` which takes cares of flattening.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterMergeRule.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+      this.exprs = exprs;
+    }
+
+    @Override
+    public RexNode visitLocalRef(RexLocalRef localRef) {
+      RexNode tree = this.exprs.get(localRef.getIndex());
+      return RexUtil.flatten(rexBuilder, tree.accept(this));
+    }
+  }
+}
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterMergeRule.java:117]: If what I said above holds then we can probably get rid of this code.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterMergeRule.java (hunk 3 above)
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterMergeRule.java]: You are right... I copied the code as-is from Calcite but I think going through the program is actually not necessary.

**Comment by jcamachor** [comment]
> I guess you plan to add a test when you commit in Calcite.
Yes, I could not repro the issue in Hive and I do not see any regressions, hence that should be sufficient. However, I will add a test to Calcite as part of the commit (assuming I can reproduce the issue over there).

**Pull request closed without merging**